engine/execution/computation/computer/computer.go

```diff
@@ -400,6 +399,182 @@ func (e *blockComputer) executeBlock(
 	return res, nil
 }
 
+// executeUserTransactions executes the user transactions in the block.
+// It queues the user transactions into a request queue and then executes them in parallel.
+func (e *blockComputer) executeUserTransactions(
+	block *entity.ExecutableBlock,
+	blockSpan otelTrace.Span,
+	database *transactionCoordinator,
+	rawCollections []*entity.CompleteCollection,
+	userTxCount int,
+) {
+	txQueue := make(chan TransactionRequest, userTxCount)
+
+	e.queueUserTransactions(
+		block.ID(),
+		block.Block.Header,
+		rawCollections,
+		txQueue,
+	)
+
+	close(txQueue)
+
+	e.executeQueue(blockSpan, database, txQueue)
+}
+
+// executeSystemTransactions executes all system transactions in the block as part of the system collection.
+//
+// System transactions are executed in the following order:
+// 1. system transaction that processes the scheduled callbacks which is a blocking transaction and
+// the result is used for the next system transaction
+// 2. system transactions that each execute a single scheduled callback by the ID obtained from events
+// of the previous system transaction
+// 3. system transaction that executes the system chunk
+//
+// An error can be returned if the process callback transaction fails. This is a fatal error.
+func (e *blockComputer) executeSystemTransactions(
+	block *entity.ExecutableBlock,
+	blockSpan otelTrace.Span,
+	database *transactionCoordinator,
+	rawCollections []*entity.CompleteCollection,
+	userTxCount int,
+) error {
+	userCollectionCount := len(rawCollections)
+	txIndex := uint32(userTxCount)
+
+	systemCtx := fvm.NewContextFromParent(
+		e.systemChunkCtx,
+		fvm.WithBlockHeader(block.Block.Header),
+		fvm.WithProtocolStateSnapshot(e.protocolState.AtBlockID(block.ID())),
+	)
+
+	systemLogger := systemCtx.Logger.With().
+		Str("block_id", block.ID().String()).
+		Uint64("height", block.Block.Header.Height).
+		Bool("system_chunk", true).
+		Bool("system_transaction", true).
+		Int("num_collections", userCollectionCount).
+		Logger()
+
+	systemCollectionInfo := collectionInfo{
+		blockId:             block.ID(),
+		blockIdStr:          block.ID().String(),
+		blockHeight:         block.Block.Header.Height,
+		collectionIndex:     len(rawCollections),
+		CompleteCollection:  &entity.CompleteCollection{},
+		isSystemTransaction: true,
+	}
+
+	var callbackTxs []*flow.TransactionBody
+	var err error
+
+	if e.vmCtx.ScheduleCallbacksEnabled {
+		callbackTxs, err = e.executeProcessCallback(
+			systemCtx,
+			systemCollectionInfo,
+			database,
+			blockSpan,
+			txIndex,
+			systemLogger,
+		)
+		if err != nil {
+			return err
+		}
+
+		txIndex++
```

Comment:
Why ++ here?

It's better an updated txIndex is returned by `executeProcessCallback`:

```
		callbackTxs, txIndexAfterExecutingProcessCallback, err = e.executeProcessCallback(...)
		...
		txIndex = txIndexAfterExecutingProcessCallback

```